orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package filerepo
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/pkg/errors"
+)
+
+const (
+	repoFilePermPrivateRW os.FileMode = 0600
+)
+
+type Repo struct {
+	mu                  sync.Mutex
+	fileRepoDir         string
+	fileSuffix          string
+	transientFileMarker string
+}
+
+type RepoOption func(r *Repo) error
+
+func WithTransientFileMarker(marker string) RepoOption {
+	return func(r *Repo) error {
+		r.transientFileMarker = marker
+		return nil
+	}
+}
+
+// New initializes a new file repo at repoParentDir/fileSuffix. It takes optional
+// RepoOption arguments for modifying the default repo options. All file system operations
+// on the returned file repo are thread safe.
+func New(repoParentDir, fileSuffix string, opts ...RepoOption) (*Repo, error) {
+	// default transientFileMarker to ~, can modify via opts
+	repo := &Repo{
+		transientFileMarker: "~",
+		fileSuffix:          fileSuffix,
+		fileRepoDir:         filepath.Join(repoParentDir, fileSuffix),
+	}
+
+	for _, opt := range opts {
+		if err := opt(repo); err != nil {
```

Comment:
I'll remove the configurability for now as I don't see an immediate need to override the default value. We can always revisit this down the line if an override is needed.

Reply:
In this instance, fine. In general, this is the wrong way.

When it comes to constructing objects, we have:

1. Zero values
2. Struct literals that require nothing more than wiring
3. Constructors that take no arguments
4. Constructors that take specific arguments
5. Constructors that can evolve

Option 1 is best. It's the "make the zero value useful" goal of of the standard library.

Option 2 is good when evolving from option 1. This allows us to create a default behavior with the zero value and add customizability after the fact by injecting something to override behavior. When we set a field, we have an explicit name at the call site so we know what we're providing. We see a lot of this in the http package.

Option 3 is okay but limiting. It means that we're constructing something and there's no way to control it. That implies we end up having to modify the state after construction to make things useful or build global state around it to influence it.

Option 4 is rigid and confusing. The rigidity comes from the fact that there's never a backward compatible way to add or remove constructor arguments without impacting callers. It's confusing when we get long lists of arguments that share the same type. In this particular case, it looks like we have `n` strings. So, at the call site, I need to figure out the order and what they mean.

Option 5 is good. In this model we can easy extend the functionality of constructors. Each option is named so when we make the call, we know what we're providing. The options allow better encapsulation of the object as even the names of the fields we use to hold dependencies is not exposed to the caller.

Given how prevalent the option func pattern is in libraries, we should all be able to recognize it for what it is and embrace the benefits it provides when it comes to evolving a code base.

Reply:
Thanks, @caod123 for removing this configuration option.

@sykesm Thanks for the nice summary. Especially what you mentioned in option 4 and 5 are nice -- no need to break backward compatibility when we add new configuration option -- didn't strike me. While optionalal func pattern might add a bit of overhead while reading it for the first time (as in my case), we will adjust to it as we see more patterns. I hope you don't mind that I am copying your suggestions to the [google doc](https://docs.google.com/document/d/1mU4S4U1dDpBk9AA180RMRB7g3OH887nqbcEPGD3px3Q/edit?usp=sharing) that I maintain so that these suggestions/recommendations are not lost within the PR comments.